docs: Golang filter requires disabling cgocheck. (#25222)
envoyproxy/envoy: docs/root/configuration/http/http_filters/golang_filter.rst

<https://docs.google.com/document/d/1noApyS0IfmOGmEOHdWk2-BOp0V37zgXMM4MdByr1lQk/edit?usp=sharing>`_
for more details on the filter's implementation.

> .. attention::

Reviewer: ```suggestion
.. warning::
```

  The Envoy Golang filter is designed to be run with the `GODEBUG=cgocheck=0` environment variable set.

>   This disables the cgo pointer check.

Reviewer: ```suggestion
  This disables the cgo pointer check.
  
  Failure to set this environment variable will cause Envoy to crash!
```

for more details on the filter's implementation.

.. attention::

>   It is designed to work with the `GODEBUG=cgocheck=0` ENV variable, when running envoy with Go plugin enabled,

Reviewer: ```suggestion
  The Envoy Golang filter is designed to work with the `GODEBUG=cgocheck=0` environment variable
```

Im not totally clear if this relates to building, running or both

Reviewer: i guess my question then is where should this be set - i can see it in the tests but not anywhere else

does this mean that an end-user _must_ do:

```console
$ export GODEBUG=cgocheck=0
$ ./bin/envoy -c path-to-config-with-go-filter.yaml
```

Author: > does this mean that an end-user _must_ do:
> 
> ```
> $ export GODEBUG=cgocheck=0
> $ ./bin/envoy -c path-to-config-with-go-filter.yaml
> ```

yep, that's expected.

Reviewer: hmm - that seems less than ideal

can we not do this in code?

what happens if the end-user forgets/omits to do this?

Author: > can we not do this in code?

Yep, do it in code should be better. It's already on our TODO list.

> what happens if the end-user forgets/omits to do this?

Go will panic, envoy will crash then.

Reviewer: > Yep, do it in code should be better. It's already on our TODO list.

hmm, well - given that it can crash we defo need to do something - happy to land this until it is fixed

im wondering tho - would it not be easier to fix than document the workaround ?

Reviewer: i guess the other ~related question (which is less of a priority) is what if a user _does_ want to debug with `cgocheck=1` (excuse my limited go knowledge if this is a stupid question)

Author: > im wondering tho - would it not be easier to fix than document the workaround ?

There may be other items in the `GODEBUG` env variable, i.e. `GODEBUG=gctrace=1,invalidptr=1,cgocheck=0,madvdontneed=1`, and we only overwriting the `cgocheck` item.
So, it is a bit complicated, but, yep, it's not a big deal.

I personally prefer doc first, and we still need a doc for hardcoding this in code when we do it in code.
It's also ok for me, if you think doing it in code first could be better.

> what if a user does want to debug with `cgocheck=1`

Oh, I think `cgocheck=1` is not for debugging.
In short, `cgocheck` checks we could not pass a struct with Go pointer inside, to C side, so that C could not touch the Go pointer - to make Go memory safe, could not be modified by C unexpectedly.
It's a higher level of protection for people who do not understand the cgo details.

Go filter extension is a cgo heavily use case, so we disable it, use low level of protection instead - as discussed here: https://github.com/envoyproxy/envoy/issues/25178#issuecomment-1404995919

Note: users who use the Go APIs from the Go filter extension, do not need to care about the `cgocheck`, Go APIs are always memory safety.

Reviewer: thanks for the explanation re GODEBUG - i guess updating requires parsing what is there already

> I personally prefer doc first, and we still need a doc for hardcoding this in code when we do it in code.

i am definitely happy getting docs with code - but this is not a simple documentation issue

The current default (ie GODEBUG _not_ set) causes Envoy to crash - this is one of the worst type of bugs - addressing that is a huge priority

Documenting broken software is generally a bad idea if fixing it is more/less trivial

>  and we still need a doc for hardcoding this in code when we do it in code.

sure, but hopefully it wont have to warn anyone about anything crashing

for more details on the filter's implementation.

.. attention::

>   The Envoy Golang filter is designed to be run with the `GODEBUG=cgocheck=0` environment variable set,
>   when running Envoy with Golang filter enabled, which disables the cgo pointer check.

Reviewer: ```suggestion
  The Envoy Golang filter is designed to be run with the `GODEBUG=cgocheck=0` environment variable set.
  
  This disables the cgo pointer check.
```

the sentence doesnt quite read correctly

Author: Thanks, fixed it with a force push, due to miss DSO in the previous commit.
